includes/profile.inc

```diff
@@ -449,3 +459,368 @@ function cm_tools_update_missing_default_field_values_insert_with_languages(arra
+ *                   │
+ *                   │  ┌───────────────────────────────────┐
+ *          move to next table     ┌─────────────────────┐  │
+ *                   │  │          │      Next chunk     │  │
+ *           ┌───────▼──▼─────┐    │  ┌────────────────┐ │  │
+ *           │ GATHER ROWS    │    └──► PROCESS ROWS   ├─┘  │
+ *           │ - <table>      ├───────► - <table>      ├────┘
+ *           │                │ First │ - <rows chunk> │
+ *           │                │ chunk └────────────────┘
+ *           └───────┬────────┘
+ *                   │ No tables
+ *                   │  left
+ *                   ▼
+ *                  END
+ */
+  // Convenience reference.
+  if (!isset($sandbox['_cm_tools_tables_walk'])) {
+    $sandbox['_cm_tools_tables_walk'] = array();
+  }
+  $s = &$sandbox['_cm_tools_tables_walk'];
+
+  $status = array();
+
+  if (!is_callable($tables)) {
+    $move_to_next_table = function () use (&$s) {
+      return array_shift($s['tables']);
+    };
+  }
+  else {
+    $move_to_next_table = $tables;
+  }
+
+  // Initialize state and shared variables.
+  if (empty($s)) {
+    $s = array(
+      'state' => 'GATHER ROWS',
+      'tables' => $tables,
+      'table' => NULL,
+      'rows' => array(),
+      'rows_updated' => 0,
+    );
+    $s['table'] = $move_to_next_table();
+  }
+
+  switch ($s['state']) {
+
+    case 'GATHER ROWS':
+      if (!isset($s['table'])) {
+        $s['state'] = 'END';
+        break;
+      }
+      list($table_name, $primary_columns) = $s['table'];
+      $query = db_select($table_name, 'f')
+        ->fields('f', $primary_columns);
+      if (is_callable($query_alter_callback)) {
+        call_user_func($query_alter_callback, $query, $table_name);
+      }
+      $s['rows'] = $query
+        ->execute()
+        ->fetchAll(PDO::FETCH_ASSOC);
+      $s['state'] = 'PROCESS ROWS';
+      $status[] = strtr('Found !row_count rows in !table_name.', array(
+        '!row_count' => count($s['rows']),
+        '!table_name' => $table_name,
+      ));
+      // Deliberately fall-through, because why not?
+      //break;
+
+    case 'PROCESS ROWS':
+      list($table_name, $primary_columns) = $s['table'];
+      $rows = array_splice($s['rows'], 0, $rows_per_pass);
+      if (empty($rows)) {
+        $status[] = strtr('Updated !row_count rows in !table_name.', array(
+          '!row_count' => $s['rows_updated'],
+          '!table_name' => $table_name,
+        ));
+        $s['rows_updated'] = 0;
+        $s['state'] = 'GATHER ROWS';
+        $s['table'] = $move_to_next_table();
+        break;
+      }
+      // Fetch the values for all the rows in our chunk in the same query.
+      $qry = db_select($table_name, 'f')
+        ->fields('f');
+      $or = db_or();
+      foreach ($rows as $row) {
+        $and = db_and();
+        foreach ($row as $primary_column => $value) {
+          $and->condition($primary_column, $value);
+        }
+        $or->condition($and);
+      }
```

Comment:
Not sure of the performance here. But if there's only a single primary column, it could at least just use a single condition with the 'IN' operator perhaps?

Reply:
Field data tables have 5 primary columns (entity_type, entity_id, deleted, delta, language) and field revision tables have 6 (same as data plus revision_id).

Reply:
Ah. Perhaps you're suggesting that we introduce a special-case in the code for when there is only 1 primary column? Yeah, I don't know how that would affect performance. I would hope mysql optimises things like that away, but I don't really know.

Perhaps best to wait until we run into a problem?

Reply:
Yeah, that was my intention. Having this code in a general-purpose tools module implies it could be used for a more general purpose. Otherwise ... I vote for moving all this into spire, as I can't see it being used beyond there.